Bit of clean up and a Dockerfile

aaronelliotross opened the pull request:
This is a batch of semi-related changes - getting ready to put this code in staging as a docker container. The diff is giant because of package-lock.json

 - Put URL path and timestamps on log entries
 - Limit CORS to configured domains
 - Add a Dockerfile
 - Add a README

tttp commented:
Hi,

I'm not sure about the CORS, it might block legitimate usage, eg if the widget is embedded in a partner website or a micro-site for the campaign "ask your boat captain to block the world economy, on ship at a time" (on evergreen.suez)

What is the risk adding CORS is trying to prevent? I'm not sure I see what we are trying to block.

If we want to pursue with explicit CORS, could you handle the path that if PROCA_DONATE_CORS_DOMAINS isn't set, we default to "*"?

On a side note, my initial plan was to run it as a serverless function (cloudflare worker), and it seems stripe has now a solution to run it there.

Could you add a .env.example with the 
   PROCA_DONATE_CORS_DOMAINS=example.org;www.example.org ? I tend to find it easier to keep the config into a .env.example rather than the readme

also, my impression is that it's more "standard" to use "," instead of ";" as the separator, but I don't have a strong opinion about it, happy to be overruled

aaronelliotross commented:
> I'm not sure about the CORS, it might block legitimate usage, eg if the widget is embedded in a partner website or a micro-site for the campaign "ask your boat captain to block the world economy, on ship at a time" (on evergreen.suez)

Once this is up and running, I'll donate to that campaign. =)
 
> What is the risk adding CORS is trying to prevent? I'm not sure I see what we are trying to block.

Setting the CORS allow origin header to '*' allows cross domain access to any site.  Since the Stripe publishable keys are public - any site could use any part of the Stripe API exposed through this service. I'm pretty sure I could use PaymentIntents to test card numbers.
 
> Could you add a .env.example with the
> PROCA_DONATE_CORS_DOMAINS=example.org;www.example.org ? I tend to find it easier to keep the config into a .env.example rather than the readme

Sure. I'll update the delimiter too.

Changes coming your way!

aaronelliotross commented on Dockerfile:
```diff
@@ -0,0 +1,13 @@
+
+FROM keymetrics/pm2:latest-slim
```
Ooops - there's no need to use the keymetrics image. I'll switch back to node:14.

aaronelliotross commented:
Oh! another thing I changed here was to get rid of the yarn.lock file - assuming everyone would rather just use node / npm. One less thing to install!

https://github.com/fixthestatusquo/proca-donate/pull/1/files#diff-51e4f558fae534656963876761c95b83b6ef5da5103c4adef6768219ed76c2de

tttp commented:
> Setting the CORS allow origin header to '*' allows cross domain access to any site. Since the Stripe publishable keys are public - any site could use any part of the Stripe API exposed through this service. I'm pretty sure I could use PaymentIntents to test card numbers.

Digging further: what's prevent me to use the service from curl or any other that doesn't care about CORS?
Still not seeing the risk, but given that the config can be set to '*' anyway, it's better to allow to change it to something more restricted.

your PR has a conflict on yarn.lock (I updated of stripe ;), can you resolve it and I'll pull?

aaronelliotross commented:
> Digging further: what's prevent me to use the service from curl or any other that doesn't care about CORS?

True - if I have a list of cards there are lots of ways of testing them. Any credit card form will do. 

But this would allow collecting cards and testing them immediately - a google ad, a fake site and I can start collecting cards using someone else's Stripe account. 

Maybe I'm mistaken, but I'm pretty sure the general risk here is letting anyone on the internet set up a site and use our Stripe account.
 
> your PR has a conflict on yarn.lock (I updated of stripe ;), can you resolve it and I'll pull?

Oops I was intentionally removing yarn. Is that OK?  I don't have a strong opinion at all - just simplifying.

tttp commented:
I'm fine with removing yarn and indeed, one less dependency

but the conflict still need to be resolved, happy to do it if you prefer ;)

aaronelliotross commented:
> 
> 
> I'm fine with removing yarn and indeed, one less dependency
> 
> but the conflict still need to be resolved, happy to do it if you prefer ;)

Done!

tttp commented:
Hi @aaronelliotross ,

FYI: I ran the install locally and the package-lock was quite different. I assume it's because you installed pm2 separately? 

I cleaned up (well, npm install did)

and I've modified npx run start so it works for sure (node only) and added for your a "npx run dev" that is using nodemon .

(I've also added nodemon and pm2 as peer dependency, not sure it's the best way to flag it can be needed/useful in some case, but the best I could think of)

aaronelliotross commented:
Ah sorry about the package lock difference - I thought a clean install would fix that!

Thanks for fixing.
On 3/26/2021 10:17:22 AM, xavier dutoit ***@***.***> wrote:
Hi @aaronelliotross [https://github.com/aaronelliotross] ,
FYI: I ran the install locally and the package-lock was quite different. I assume it's because you installed pm2 separately?
I cleaned up (well, npm install did)
and I've modified npx run start so it works for sure (node only) and added for your a "npx run dev" that is using nodemon .
(I've also added nodemon and pm2 as peer dependency, not sure it's the best way to flag it can be needed/useful in some case, but the best I could think of)
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub [https://github.com/fixthestatusquo/proca-donate/pull/1#issuecomment-808061792], or unsubscribe [https://github.com/notifications/unsubscribe-auth/AAA4RTQPE2AYRLCLGSWEXDLTFRGJ7ANCNFSM4ZZMI7VQ].
